engine/table/src/main/java/io/deephaven/engine/table/impl/sources/ring/AbstractRingChunkSource.java

```diff
@@ -0,0 +1,440 @@
+    }
+
+    /**
+     * Equivalent to {@code append(source, srcKeys, APPEND_CHUNK_SIZE)}.
+     *
+     * <p>
+     * A generic append with a conservatively sized {@code appendChunkSize}. If the caller knows that the {@code source}
+     * uses a {@link FillContext} with {@link FillContext#supportsUnboundedFill()}, they should instead call
+     * {@link #appendUnbounded(ChunkSource, RowSet)}.
+     * 
+     * @param source the source
+     * @param srcKeys the source keys
+     * @see #append(ChunkSource, RowSet, int)
+     */
+    public final void appendBounded(ChunkSource<? extends Values> source, RowSet srcKeys) {
+        append(source, srcKeys, APPEND_CHUNK_SIZE);
+    }
+
+    /**
+     * Equivalent to {@code append(source, srcKeys, capacity)}.
+     *
+     * <p>
+     * A specialized append, where the caller knows that {@code source} is using a {@link FillContext} with
+     * {@link FillContext#supportsUnboundedFill()}.
+     *
+     * @param source the source
+     * @param srcKeys the source keys
+     * @see #append(ChunkSource, RowSet, int)
+     */
+    public final void appendUnbounded(ChunkSource<? extends Values> source, RowSet srcKeys) {
+        append(source, srcKeys, capacity);
+    }
+
+    /**
+     * Append the data represented by {@code source} and {@code srcKeys} into {@code this} ring. This method is meant to
+     * be efficient, and will read at most {@link #capacity()} items from the end of {@code source} and {@code srcKeys}.
+     * The {@link #lastKey() lastKey} will increase by {@code srcKeys.size()}.
+     *
+     * @param source the source
+     * @param srcKeys the source keys
+     * @param appendChunkSize the append chunk size
+     */
+    public final void append(ChunkSource<? extends Values> source, RowSet srcKeys, int appendChunkSize) {
+        if (srcKeys.isEmpty()) {
+            return;
+        }
+        final long logicalFillSize = srcKeys.size();
+        final RowSet physicalRows;
+        final long physicalStartRingIx;
+        final boolean hasSkippedRows = logicalFillSize > capacity;
+        if (!hasSkippedRows) {
+            physicalRows = srcKeys;
+            physicalStartRingIx = nextRingIx;
+        } else {
+            final long skipRows = logicalFillSize - capacity;
+            physicalRows = srcKeys.subSetByPositionRange(skipRows, logicalFillSize);
+            physicalStartRingIx = nextRingIx + skipRows;
+        }
+        final int fillStartIx = keyToRingIndex(physicalStartRingIx);
+        try (
+                final RowSet _physicalRowsToClose = hasSkippedRows ? physicalRows : null;
+                final ResettableWritableChunk<Any> chunk = getChunkType().makeResettableWritableChunk();
+                final FillContext fillContext = source.makeFillContext(appendChunkSize);
+                final Iterator it = physicalRows.getRowSequenceIterator()) {
+            // Note: if we could do it.skip(skipRows), that would probably be more efficient than subSetByPositionRange
```

Comment:
We discussed. Better to not dot he `subsetByPositionRange()`. Instead do `it.advance(srcKeys.get(skipRows))`. If you want to be fancy, you could have an assertion around `advanceAndGetPositionDistance()`.

Reply:
Cool.